### ISSUE TITLE
Computation of n_levels in ActUniformQuantizer for symmetric quantisation.

### ISSUE THREAD

**Issue by adilhasan927** [hatchetProject/QuEST#16]
Hello,

In the q-diffusion repository, we have:

```
    def __init__(self, n_bits: int = 8, symmetric: bool = False, channel_wise: bool = False, scale_method: str = 'max',
                 leaf_param: bool = False, always_zero: bool = False):
        super(UniformAffineQuantizer, self).__init__()
        self.sym = symmetric
        # assert 2 <= n_bits <= 8, 'bitwidth not supported'
        self.n_bits = n_bits
        self.n_levels = 2 ** self.n_bits if not self.sym else 2 ** (self.n_bits - 1) - 1
        ...

    def forward(self, x: torch.Tensor):
        ....
        if self.sym:
            x_quant = torch.clamp(x_int, -self.n_levels - 1, self.n_levels)
        else:
            x_quant = torch.clamp(x_int, 0, self.n_levels - 1)
```

I would like to draw attention to: **self.n_levels = 2 ** self.n_bits if not self.sym else 2 ** (self.n_bits - 1) - 1**. This has special logic to calculate the number of levels which can be encoded by the number of bits when using symmetric quantisation, due to the necessity to consider negative values under 2s complement arithmetic.

As an example: If I have a **uint4**, I can represent values **0, 1,** ... 2^4-1=**15**, corresponding to 16 distinct levels.

But, if I want to represent negative values on the left hand of zero as well, then under 2's complement arithmetic, now my 16 distinct levels become -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, again corresponding to 16 distinct levels, of which 7 are positive.

These two cases of non-symmetric vs symmetric are what are correctly handled in the Q-diffusion codebase in the init() method.

I would now like to draw attention to this code in **ActUniformQuantizer**:

```
    def __init__(self, n_bits: int = 8, symmetric: bool = False, channel_wise: bool = False, scale_method: str = 'max',
                 leaf_param: bool = False, always_zero: bool = False):
        super(ActUniformQuantizer, self).__init__()
        # assert 2 <= n_bits <= 8, 'bitwidth not supported'
        self.sym = symmetric
        self.always_zero = always_zero
        self.n_bits = n_bits
        self.n_levels = 2 ** self.n_bits
        ....

    def forward(self, x: torch.Tensor):
        if self.sym:
            x_quant = torch.clamp(x_int, -self.n_levels - 1, self.n_levels)
        else:
            x_quant = torch.clamp(x_int, 0, self.n_levels - 1)
```

As we can see, the clamping in forward() is identical to Xiuyu Li's. However, the check for self.sym in the init() method is not present, and instead we have  **`self.n_levels = 2 ** self.n_bits`** *even in the symmetric case*.

Thus, now if we have n_bits=4, sym=True, the resulting rounded and clamped ranged is: -17, -16, -15, -14, -13, ..., -1, 0, 1, 13, 14, 15, 16, for a total of 34 distinct values, even more than is representable with 5 bits (i.e. if we add a 5th bit to serve as sign bit)...

Is this an intentional characteristic of the quantisation method/am I misunderstanding the code?

**Comment by hatchetProject** [comment]
Hi, thank you very much for pointing this out. Your understanding is correct. I **sincerely apologize** for this mistake during code upload. Fortunately, only the church dataset for this repo is influenced. As for our results in the paper, we checked closely about the implementation and they are correct.

The mistake is due to the fact while uploading code, I used the quant_layer.py file of SD, which always sets `self.sys = False`. We practically used different folders for each task and during combination and there are some inconsistencies, where the line of code is always `self.n_levels = 2 ** self.n_bits if not self.sym else 2 ** (self.n_bits - 1) - 1` in `__init__()` for the other tasks than SD. 

I will update the repo also. Apologies again for the confusion.
